frost-core/src/frost.rs

```diff
@@ -64,19 +65,61 @@ where
     }
 }
 
-impl<C> From<&SigningPackage<C>> for Rho<C>
+/// A list of binding factors and their associated identifiers.
+#[derive(Clone)]
+pub struct BindingFactorList<C: Ciphersuite>(Vec<(Identifier<C>, Rho<C>)>);
+
+impl<C> BindingFactorList<C>
 where
     C: Ciphersuite,
 {
-    // [`compute_binding_factor`] in the spec
-    //
-    // [`compute_binding_factor`]: https://www.ietf.org/archive/id/draft-irtf-cfrg-frost-03.html#section-4.4
-    fn from(signing_package: &SigningPackage<C>) -> Rho<C> {
-        let preimage = signing_package.rho_preimage();
+    /// Return iterator through all factors.
+    pub fn iter(&self) -> impl Iterator<Item = &(Identifier<C>, Rho<C>)> {
+        self.0.iter()
+    }
+}
+
+impl<C> Index<Identifier<C>> for BindingFactorList<C>
+where
+    C: Ciphersuite,
+{
+    type Output = Rho<C>;
 
-        let binding_factor = C::H1(&preimage[..]);
+    // Get the binding factor of a participant in the list.
+    //
+    // [`binding_factor_for_participant`] in the spec
+    //
+    // [`binding_factor_for_participant`]: https://www.ietf.org/archive/id/draft-irtf-cfrg-frost-07.html#section-4.3
+    fn index(&self, identifier: Identifier<C>) -> &Self::Output {
```

Comment:
```suggestion
    // TODO: switch from Vec to BTreeMap, as this can be made more efficient.
    fn index(&self, identifier: Identifier<C>) -> &Self::Output {
```